File: implementation-nodejs/user-service/src/handler.js
const grpc = require('@grpc/grpc-js');
const userService = require('./service');

module.exports = {
    validateUser: async (call, callback) => {
        try 
        {
            const { user_id, amount_to_spend } = call.request;

            if(user_id <= 0) 
            {
                return callback({
                    code: grpc.status.INVALID_ARGUMENT,
                    details: "Invalid User ID"
                });
            }

            if(amount_to_spend <= 0) 
            {
                return callback({
                    code: grpc.status.INVALID_ARGUMENT,
                    details: "Invalid amount to spend"
                })
            }

            const result = await userService.checkEligibility(user_id, amount_to_spend);
            return callback(null, {
                is_valid: result.isValid,
                name: result.name || "",
                balance_sufficient: result.balanceSufficient,
                current_balance: result.currentBalance
            });
        }
        catch(error)
        {
            callback({ code: grpc.status.INTERNAL, details: error.message });
        }
    },
    updateUserBalance: async (call, callback) => {
        try
        {
            const { user_id, amount, is_credit } = call.request;

            if(user_id <= 0) 
            {
                return callback({
                    code: grpc.status.INVALID_ARGUMENT,
                    details: "Invalid User ID"
                });
            }

            if(amount <= 0) 
            {
                return callback({
                    code: grpc.status.INVALID_ARGUMENT,
                    details: "Invalid amount"
                })
            }

            const result = await userService.updateUserBalance(user_id, amount, is_credit);
            return callback(null, {
                status: result.status,
                message: result.message,
                current_balance: result.balance
            });
        }
        catch(error)
        {
            callback({ code: grpc.status.INTERNAL, details: error.message });
        }
    }
}

File: implementation-go/user-service/handler.go
package user_service

import (
	"context"
	"go-microtrans-grpc/user-service/pb/user"

	"google.golang.org/grpc/codes"
	"google.golang.org/grpc/status"
)

type UserHandler struct {
	user.UnimplementedUserServiceServer
	service *UserService
}

func (h *UserHandler) ValidateUser(ctx context.Context, req *user.UserEligibilityRequest) (*user.UserEligibilityResponse, error) {
	userId := req.GetUserId()
	amountToSpend := req.GetAmountToSpend()

	if userId <= 0 {
		return nil, status.Error(codes.InvalidArgument, "Invalid User ID")
	}

	if amountToSpend <= 0 {
		return nil, status.Error(codes.InvalidArgument, "Invalid amount to spend")
	}

	targetUser, eligible, _ := h.service.CheckUserEligibility(userId, amountToSpend)
	if targetUser == nil {
		return &user.UserEligibilityResponse{
			IsValid: false,
		}, nil
	}

	return &user.UserEligibilityResponse{
		IsValid:           true,
		Name:              targetUser.Name,
		CurrentBalance:    targetUser.Balance,
		BalanceSufficient: eligible,
	}, nil
}

func (h *UserHandler) UpdateUserBalance(ctx context.Context, req *user.UpdateBalanceRequest) (*user.UpdateBalanceResponse, error) {
	userId := req.GetUserId()
	amount := req.GetAmount()
	isCredit := req.GetIsCredit()

	if userId <= 0 {
		return nil, status.Error(codes.InvalidArgument, "Invalid User ID")
	}

	if amount <= 0 {
		return nil, status.Error(codes.InvalidArgument, "Invalid amount")
	}

	status, message, newBalance := h.service.UpdateUserBalance(userId, amount, isCredit)
	return &user.UpdateBalanceResponse{
		Status:         status,
		Message:        message,
		CurrentBalance: newBalance,
	}, nil
}

func NewUserHandler() *UserHandler {
	return &UserHandler{
		service: NewUserService(),
	}
}
